cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java

```diff
@@ -163,6 +184,70 @@ public OperationSnapshot createCluster(String name) throws RetryableProvisionExc
       // otherwise, it's not a retryable failure
       throw e;
     }
+
+    List<Node> nodes = new ArrayList<>();
+    for (String masterName : cluster.getConfig().getMasterConfig().getInstanceNamesList()) {
+      nodes.add(getNode(compute, Constants.Node.MASTER_TYPE, masterName));
+    }
+    for (String workerName : cluster.getConfig().getWorkerConfig().getInstanceNamesList()) {
+      nodes.add(getNode(compute, Constants.Node.WORKER_TYPE, workerName));
+    }
+    return Optional.of(new co.cask.cdap.runtime.spi.provisioner.Cluster(
+      cluster.getClusterName(), convertStatus(cluster.getStatus()), nodes, Collections.emptyMap()));
+  }
+
+  private Node getNode(Compute compute, String type, String nodeName) throws IOException {
+    Instance instance;
+    try {
+      instance = compute.instances().get(conf.getProjectId(), conf.getZone(), nodeName).execute();
+    } catch (GoogleJsonResponseException e) {
+      // this can happen right after a cluster is created
+      if (e.getStatusCode() == 404) {
+        return new Node(nodeName, -1L, Collections.emptyMap());
```

Comment:
why not retry? Otherwise, we won't have some necessary information, right?

Reply:
By the time the cluster gets to the running state, this won't happen, and the cluster seen at that time is what will get passed on. Since this method is also used to poll for status, I didn't want to make it a long operation.

Reply:
Ok, that guarantee wasn't apparent, but then that's fine.